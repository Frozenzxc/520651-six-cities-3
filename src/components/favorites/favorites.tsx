import * as React from "react";
import {getCitiesList} from "../../utils";
import PlaceCard from "../place-card/place-card";
import {getFavoriteOffers} from "../../reducer/offers/selectors";
import {connect} from "react-redux";
import NameSpace from "../../reducer/name-space";
import {Operation} from "../../reducer/offers/offers";
import FavoritesEmpty from "../favorites-empty/favorites-empty";
import withPlaceCardComponent from "../../hocs/with-place-card-component/with-place-card-component";
import {Offer} from "../../types";

interface Props {
  addToFavorite: () => void;
  authEmail: string;
  authorizationStatus: string | null;
  isFavoritesLoading: boolean;
  loadFavoriteOffers: () => void;
  favoriteOffers: Offer[];
  onCardHover: () => void;
  onCardTitleClick: () => void;
}

const FavoriteOfferType = `favorites`;

const PlaceCardWrapped = withPlaceCardComponent(PlaceCard);

class Favorites extends React.PureComponent<Props, {}> {
  componentDidMount() {
    const {loadFavoriteOffers} = this.props;
    loadFavoriteOffers();
  }

  render() {
    const {authorizationStatus, addToFavorite, authEmail, isFavoritesLoading, favoriteOffers, onCardHover, onCardTitleClick} = this.props;


    if (isFavoritesLoading) {
      return false;
    }

    if (!favoriteOffers.length) {
      return <FavoritesEmpty authorizationStatus={authorizationStatus} authEmail={authEmail}/>;
    }

    const citiesList = getCitiesList(favoriteOffers);
    return (
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {citiesList.map((city) => {
                return (
                  <li className="favorites__locations-items" key={city}>
                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link" href="#">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>
                    <div className="favorites__places">
                      {favoriteOffers.filter((offer) => offer.city.name === city).map((card) => {
                        return (
                          <PlaceCardWrapped
                            authorizationStatus={authorizationStatus}
                            addToFavorite={addToFavorite}
                            onMouseEnter={onCardHover}
                            onClick={onCardTitleClick}
                            offer={card}
                            key={card.id}
                            offersView={FavoriteOfferType}
                          />);
                      })}
                    </div>
                  </li>);
              }
              )}
            </ul>
          </section>
        </div>
      </main>
    );
  }
}

const mapStateToProps = (state) => ({
  authEmail: state[NameSpace.USER].authEmail,
  authorizationStatus: state[NameSpace.USER].authorizationStatus,
  isFavoritesLoading: state[NameSpace.OFFERS].isFavoritesLoading,
  favoriteOffers: getFavoriteOffers(state),
});

const mapDispatchToProps = (dispatch) => ({
  addToFavorite(offer) {
    dispatch(Operation.addToFavorite(offer));
  },
  loadFavoriteOffers() {
    dispatch(Operation.loadFavoriteOffers());
  },
});

export {Favorites};
export default connect(mapStateToProps, mapDispatchToProps)(Favorites);

