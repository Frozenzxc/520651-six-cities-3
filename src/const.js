const OfferType = {
  ALL: `cities`,
  NEARBY: `near`,
};

const SortType = {
  POPULAR: `Popular`,
  PRICE_TO_HIGH: `Price: low to high`,
  PRICE_TO_LOW: `Price: high to low`,
  RATING: `Top rated first`,
};

const MAX_NEARBY_OFFERS = 3;

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const ReviewPostingStatus = {
  POSTED: `POSTED`,
  ERROR: `ERROR`,
};

export {OfferType, SortType, MAX_NEARBY_OFFERS, AuthorizationStatus, ReviewPostingStatus};
