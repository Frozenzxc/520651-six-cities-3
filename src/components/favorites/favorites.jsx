import React, {PureComponent} from "react";
import {offerShape} from "../../prop-types.jsx";
import PropTypes from "prop-types";
import {getCitiesList} from "../../utils";
import PlaceCard from "../place-card/place-card.jsx";
import {OfferType} from "../../const";
import {getFavoriteOffers} from "../../reducer/offers/selectors";
import {connect} from "react-redux";
import NameSpace from "../../reducer/name-space";
import {Operation} from "../../reducer/offers/offers";
import FavoritesEmpty from "../favorites-empty/favorites-empty.jsx";
import withPlaceCardComponent from "../../hocs/with-place-card-component/with-place-card-component";

const PlaceCardWrapped = withPlaceCardComponent(PlaceCard);

class Favorites extends PureComponent {
  componentDidMount() {
    const {loadFavoriteOffers} = this.props;
    loadFavoriteOffers();
  }

  render() {
    const {authorizationStatus, addToFavorite, authEmail, isFavoritesLoading, favoriteOffers, onCardHover, onCardTitleClick} = this.props;


    if (isFavoritesLoading) {
      return false;
    }

    if (!favoriteOffers.length) {
      return <FavoritesEmpty authorizationStatus={authorizationStatus} authEmail={authEmail}/>;
    }

    const citiesList = getCitiesList(favoriteOffers);
    return (
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {citiesList.map((city) => {
                return (
                  <li className="favorites__locations-items" key={city}>
                    <div className="favorites__locations locations locations--current">
                      <div className="locations__item">
                        <a className="locations__item-link" href="#">
                          <span>{city}</span>
                        </a>
                      </div>
                    </div>
                    <div className="favorites__places">
                      {favoriteOffers.filter((offer) => offer.city.name === city).map((card) => {
                        return (
                          <PlaceCardWrapped
                            authorizationStatus={authorizationStatus}
                            addToFavorite={addToFavorite}
                            onMouseEnter={onCardHover}
                            onClick={onCardTitleClick}
                            offer={card}
                            key={card.id}
                            offersView={OfferType.FAVORITES}
                          />);
                      })}
                    </div>
                  </li>);
              }
              )}
            </ul>
          </section>
        </div>
      </main>
    );
  }
}

Favorites.propTypes = {
  addToFavorite: PropTypes.func.isRequired,
  authEmail: PropTypes.string,
  authorizationStatus: PropTypes.string.isRequired,
  isFavoritesLoading: PropTypes.bool.isRequired,
  loadFavoriteOffers: PropTypes.func.isRequired,
  favoriteOffers: PropTypes.arrayOf(offerShape),
  onCardHover: PropTypes.func.isRequired,
  onCardTitleClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  authEmail: state[NameSpace.USER].authEmail,
  authorizationStatus: state[NameSpace.USER].authorizationStatus,
  isFavoritesLoading: state[NameSpace.OFFERS].isFavoritesLoading,
  favoriteOffers: getFavoriteOffers(state),
});

const mapDispatchToProps = (dispatch) => ({
  addToFavorite(offer) {
    dispatch(Operation.addToFavorite(offer));
  },
  loadFavoriteOffers() {
    dispatch(Operation.loadFavoriteOffers());
  },
});

export {Favorites};
export default connect(mapStateToProps, mapDispatchToProps)(Favorites);

