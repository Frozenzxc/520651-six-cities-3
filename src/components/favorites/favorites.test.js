import React from "react";
import renderer from "react-test-renderer";
import {Favorites} from "./favorites";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space";
import {AuthorizationStatus} from "../../const";
import favoriteOffers from "../../test-mocks/test-favorite-offers";

const mockStore = configureStore([]);

it(`Should Favorites component render correctly`, () => {

  const store = mockStore({
    [NameSpace.USER]: {
      authorizationStatus: AuthorizationStatus.AUTH,
      authEmail: `AAA@adfg.ru`,
    },
    [NameSpace.OFFERS]: {
      favoriteOffers,
      isFavoritesLoading: true,
    }
  });

  const tree = renderer.create(
      <Provider store={store}>
        <Favorites
          favoriteOffers={favoriteOffers}
          loadFavoriteOffers={() => {}}
          isFavoritesLoading={false}
          authorizationStatus={AuthorizationStatus.AUTH}
          authEmail={`AAA@adfg.ru`}
        />
      </Provider>
  )
        .toJSON();

  expect(tree).toMatchSnapshot();
});
