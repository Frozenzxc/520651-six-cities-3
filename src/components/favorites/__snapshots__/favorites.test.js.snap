// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Favorites component render correctly 1`] = `
<main
  className="page__main page__main--favorites"
>
  <div
    className="page__favorites-container container"
  >
    <section
      className="favorites"
    >
      <h1
        className="favorites__title"
      >
        Saved listing
      </h1>
      <ul
        className="favorites__list"
      >
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
              >
                <span>
                  Cologne
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/19.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      160
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "96%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    The house among olive 
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/11.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      192
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "84%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    Waterfront with extraordinary view
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  room
                </p>
              </div>
            </article>
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/7.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      196
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "66%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    Loft Studio in the Central Area
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  house
                </p>
              </div>
            </article>
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
              >
                <span>
                  Paris
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/7.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      350
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "88%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    Nice, cozy, warm big bed apartment
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  hotel
                </p>
              </div>
            </article>
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
              >
                <span>
                  Hamburg
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/7.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      116
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "84%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    Penthouse, 4-5 rooms + 5 balconies
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/19.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      136
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "100%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    Waterfront with extraordinary view
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
          </div>
        </li>
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
              >
                <span>
                  Brussels
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="favorites-places__card place-card"
              onMouseEnter={[Function]}
            >
              
              <div
                className="favorites-image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="110"
                    src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/6.jpg"
                    width="150"
                  />
                </a>
              </div>
              <div
                className="favorites__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      892
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "60%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                  onClick={[Function]}
                >
                  <a
                    href="#"
                  >
                    Loft Studio in the Central Area
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  house
                </p>
              </div>
            </article>
          </div>
        </li>
      </ul>
    </section>
  </div>
</main>
`;
